aten/src/ATen/native/cuda/fused_adam_utils.cuh

```diff
@@ -166,6 +173,85 @@ struct FusedAdamMathFunctor {
         }
     }
 };
-} // namespace
 
+using device_tensor_map_t = std::unordered_map<at::Device, at::Tensor>;
+device_tensor_map_t inline init_map(const c10::optional<at::Tensor>& tensor) {
+  if (!tensor.has_value()) {
+    return {};
+  } else {
+    return {{tensor->device(), tensor.value()}};
+  }
+}
+
+c10::optional<at::Tensor> inline get_device_tensor(
+    device_tensor_map_t& device_tensor_map, const c10::optional<at::Tensor>& init, const at::Device& device) {
+  if (!init.has_value()) {
+    return c10::optional<at::Tensor>();
+  } else {
+    if (!device_tensor_map.count(device)) {
+      device_tensor_map[device] = init->to(device);
+    }
+    return device_tensor_map[device];
+  }
+}
+
+// TODO(crcrpar): Make this generic so that we can utilize in foreach ops (or even expose to `torch`?)
+using nested_tensorvec_t = std::vector<std::vector<at::Tensor>>;
+using scalartype_nested_tensorvec_map_t = std::unordered_map<at::ScalarType, nested_tensorvec_t>;
+using device_scalartype_nested_tensorvec_map_t = std::unordered_map<at::Device, scalartype_nested_tensorvec_map_t>;
+device_scalartype_nested_tensorvec_map_t group_tensors_by_device_and_scalartype(
+    const std::vector<TensorList>& nested_tensorlists, const bool has_state_steps) {
+  TORCH_CHECK_GT(nested_tensorlists.size(), 0);
+  const auto num_lists{nested_tensorlists.size()};
+  const auto num_tensors{nested_tensorlists[0].size()};
+  for (const auto & i : c10::irange(num_lists)) {
+    TORCH_CHECK_EQ(num_tensors, nested_tensorlists[i].size());
+  }
+  device_scalartype_nested_tensorvec_map_t grouped_tensors;
+  constexpr auto is_tensor_okay = [](const Tensor& ref, const Tensor& t) {
+    return
+      t.is_cuda() &&
+      ref.scalar_type() == t.scalar_type() &&
+      ref.device() == t.device() &&
+      t.layout() == at::kStrided &&
+      t.is_non_overlapping_and_dense() &&
+      ref.sizes() == t.sizes() &&
+      ref.strides() == t.strides();
+  };
+  for (const auto & tensor_index : c10::irange(num_tensors)) {
+    const auto & first_tensor = nested_tensorlists[0][tensor_index];
+    const auto device = nested_tensorlists[0][tensor_index].device();
+    const auto scalar_type = nested_tensorlists[0][tensor_index].scalar_type();
+    for (const auto & i : c10::irange(num_lists)) {
+      const auto only_device_check = has_state_steps && i == num_lists - 1;
+      const auto & t = nested_tensorlists[i][tensor_index];
+      if (only_device_check) {
+        TORCH_CHECK(t.is_cuda() && device == t.device());
```

Comment:
We'd probably need a special mechanism for excepting step tensors in the generic function

Reply:
Ah, is it fair to say that in the generic grouping, we will just skip this for loop check? In other words, this check is only applicable for optimizer steps?